firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -235,31 +231,50 @@ private void persistFid(String fid) throws FirebaseInstallationsException {
   }
 
   /**
-   * Registers the FID with FIS servers if FID is in UNREGISTERED state.
+   * Registers the FID with FIS servers if FID is in UNREGISTERED state. Also, refreshes auth token
+   * if expired.
    *
    * <p>Updates FID registration status to PENDING to avoid multiple network calls to FIS Servers.
    */
   private Task<String> registerFidIfNecessary(
       PersistedFidEntry persistedFidEntry, AwaitListener listener) {
     String fid = persistedFidEntry.getFirebaseInstallationId();
 
-    // Check if the fid is unregistered
-    if (persistedFidEntry.getRegistrationStatus() == RegistrationStatus.UNREGISTERED) {
-      updatePersistedFidWithPendingStatus(fid);
+    // If FID registration is in complete/pending state or FID auth token is valid, update the
+    // listener if awaiting and return the persisted fid.
+    if (persistedFidEntry.getRegistrationStatus() != RegistrationStatus.UNREGISTERED
+        && !isAuthTokenExpired(persistedFidEntry)) {
+      return updateAwaitListenerIfRegisteredFid(persistedFidEntry, listener);
+    }
+
+    // Update FID registration status to PENDING to avoid multiple network calls to FIS Servers.
+    persistedFid.insertOrUpdatePersistedFidEntry(
+        persistedFidEntry.toBuilder().setRegistrationStatus(RegistrationStatus.PENDING).build());
```

Comment:
you need to keep "reading cache status" and "updating cache status" in one atomic section, to avoid duplicated network calls.

Reply:
Updated. PTAL.